go/store/blobstore/oci.go

```diff
@@ -0,0 +1,796 @@
+				close(completedPartChan)
+				return nil, err
+			}
+
+			batchSize = 0
+			batch = make([]*toUpload, 0)
+			continue
+		}
+
+		buf := make([]byte, partSize)
+		n, err := reader.Read(buf)
+		if err != nil {
+			if err == io.EOF {
+				break
+			}
+			close(completedPartChan)
+			return nil, err
+		}
+
+		buf = buf[:n]
+		batchSize += int64(n)
+		batch = append(batch, &toUpload{
+			b:       buf,
+			partNum: partNum,
+		})
+
+		partNum++
+	}
+
+	if batchSize > 0 && len(batch) > 0 {
+		err := uploadBatch(ctx, objectName, uploadID, batch, completedPartChan, concurrentListeners, uploadF)
+		if err != nil {
+			close(completedPartChan)
+			return nil, err
+		}
+	}
+
+	close(completedPartChan)
+
+	idx := 0
+	for cp := range completedPartChan {
+		completedParts[idx] = cp
+		idx++
+	}
+
+	sort.Slice(completedParts, func(i, j int) bool {
+		if completedParts[i].PartNum == nil || completedParts[j].PartNum == nil {
+			return false
+		}
+		return *completedParts[i].PartNum < *completedParts[j].PartNum
+	})
+	return completedParts, nil
+}
+
+func uploadBatch(ctx context.Context, objectName, uploadID string, batch []*toUpload, completedPartsChan chan objectstorage.CommitMultipartUploadPartDetails, concurrentListeners int, uploadF uploadFunc) error {
+	batchChan := make(chan *toUpload, len(batch))
+	eg, egCtx := errgroup.WithContext(ctx)
+
+	for i := 0; i < concurrentListeners; i++ {
+		eg.Go(func() error {
+			for {
+				select {
+				case <-egCtx.Done():
+					return nil
+				case u, ok := <-batchChan:
+					if !ok {
+						return nil
+					}
+
+					cp, err := uploadF(egCtx, objectName, uploadID, u.partNum, int64(len(u.b)), bytes.NewReader(u.b))
+					if err != nil {
+						return err
+					}
+
+					completedPartsChan <- cp
+				}
+			}
+		})
+	}
+
+	eg.Go(func() error {
+		defer close(batchChan)
+
+		for {
+			select {
+			case <-egCtx.Done():
+				return nil
+			default:
+				for _, u := range batch {
+					batchChan <- u
+				}
+				return nil
+			}
+		}
+	})
+
+	return eg.Wait()
+}
+
+func getDownloadInfo(br BlobRange, contentLength *int64) (int64, int64, int64, error) {
+	totalSize := int64(0)
+	if contentLength != nil {
+		totalSize = *contentLength
+	}
+
+	if totalSize == 0 {
+		return 0, 0, 0, errors.New("no content to download")
+	}
+
+	offSet := int64(0)
+	requestedSize := int64(0)
+
+	if br.offset >= totalSize {
+		offSet = totalSize - 1
+		requestedSize = 0
+	} else if br.isAllRange() {
+		requestedSize = totalSize
+	} else if br.length == 0 && br.offset != 0 {
+		if br.offset < 0 {
+			requestedSize = int64(math.Abs(float64(br.offset)))
+			offSet = totalSize + br.offset
+		} else {
+			requestedSize = totalSize - br.offset
+			offSet = br.offset
+		}
+	} else {
+		requestedSize = br.length
+		if br.offset < 0 {
+			offSet = totalSize + br.offset
+		} else {
+			offSet = br.offset
+		}
+	}
+
+	return requestedSize, totalSize, offSet, nil
+}
+
+func getUploadInfo(partSize, maxPartNum int, r io.Reader) (int, int64, io.Reader, error) {
```

Comment:
This buffers the entire file into memory. I don't think we want to do this. We either know the size of the upload or we can get it by checking for like a Size() method on the Reader, or we can just go forward with a conservative estimate on the chunk size.

In this case, we do always know the size, I think...CopyTableFile is given it explicitly, and Persist is given the memtable in memory, so it's already a bytes.Buffer. Whether that's easy to thread through to this blobstore, I'm not sure, but we already made an entirely new table persister for this...